_posts/2020-08-26-enforce-custom-resource-policies-with-opa-gatekeeper.md

```diff
@@ -0,0 +1,435 @@
+We can use the following constraint template for it:
+
+```yaml
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: strimzireplicas
+spec:
+  crd:
+    spec:
+      names:
+        kind: StrimziReplicas
+        listKind: StrimziReplicasList
+        plural: StrimziReplicas
+        singular: StrimziReplicas
+      validation:
+        openAPIV3Schema:
+          properties:
+            min-replicas:
+              type: integer
+            min-insync-replicas:
+              type: integer
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package strimzireplicas
+        violation[{"msg": msg}] {
+          not has_enough
+          msg := sprintf("Topic must have at least %v replicas and %v min.insync.replicas", [input.parameters["min-replicas"], input.parameters["min-insync-replicas"]])
+        }
+        
+        has_enough {
+          has_enough_replicas
+          has_enough_insync_replicas
+        }
+        
+        has_enough_replicas {
+          input.review.object.spec["replicas"] >= input.parameters["min-replicas"]
+        }
+        
+        has_enough_insync_replicas {
+          input.review.object.spec.config["min.insync.replicas"] >= input.parameters["min-insync-replicas"]
+        }
+```
+
+This policy has two parameters:
+* The minimal allowed number of replicas
+* The minimal allowed number of in-sync replicas
+
+And we can use these to create different constraints for different environments.
+For example, for test environment, it is ok to have only 2 replicas and `min.insync.replicas` set to 1.
+While in production, you need to use at least 3 replicas and at least 2 `min.insync.replicas`:
+
+```yaml
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: StrimziReplicas
+metadata:
+  name: strimzi-replicas
+spec:
+  match:
+    kinds:
+      - apiGroups: ["kafka.strimzi.io"]
+        kinds: ["KafkaTopic"]
+  parameters:
+    min-replicas: 3
+    min-insync-replicas: 2
+```
+
+When now someone tries to create a topic not matching the criteria, it will not be allowed:
+
+```bash
+cat <<EOF | kubectl apply -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 1
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+EOF
+Error from server ([denied by strimzi-replicas] Topic must have at least 3 replicas and 2 min.insync.replicas): error when creating "STDIN": admission webhook "validation.gatekeeper.sh" denied the request: [denied by strimzi-replicas] Topic must have at least 3 replicas and 2 min.insync.replicas
+```
+
+_Note: Gatekeeper only controls the topics created in Kubernetes using the `KafkaTopic` custom resource._
+_It does not validate topics created directly inside Kafka using Kafka APIs._
```

Comment:
That's of course true, but does it mean that we will see "errors" in the topic operator log that is trying to create the corresponding `KafkaTopic` violating the policy, right?

Reply:
Yes, that is correct. The policy would apply to the Topic Operator as well.